e2e/pipe_test.go

```diff
@@ -0,0 +1,470 @@
+}
+
+func TestUploadStdinToS3WithNoSuchUploadRetryCount(t *testing.T) {
+	t.Parallel()
+
+	s3client, s5cmd := setup(t)
+
+	bucket := s3BucketFromTestName(t)
+	createBucket(t, s3client, bucket)
+
+	const (
+		filename = "example.txt"
+		content  = "Some example text"
+	)
+
+	reader := bytes.NewBufferString(content)
+	dstpath := fmt.Sprintf("s3://%v/%v", bucket, filename)
+
+	cmd := s5cmd("pipe", "--no-such-upload-retry-count", "5", dstpath)
+	result := icmd.RunCmd(cmd, icmd.WithStdin(reader))
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: suffix(`pipe %v`, dstpath),
+	})
+
+	// assert S3
+	assert.Assert(t, ensureS3Object(s3client, bucket, filename, content))
+}
+
+// pipe --raw s3://bucket/object
+func TestUploadStdinToS3WithRawMode(t *testing.T) {
+	t.Parallel()
+
+	s3client, s5cmd := setup(t)
+
+	bucket := s3BucketFromTestName(t)
+	createBucket(t, s3client, bucket)
+
+	const (
+		filename = "ex*mple.txt"
+		content  = "Some *** text"
+	)
+
+	reader := bytes.NewBufferString(content)
+	dstpath := fmt.Sprintf("s3://%v/%v", bucket, filename)
+
+	cmd := s5cmd("pipe", "--raw", dstpath)
+	result := icmd.RunCmd(cmd, icmd.WithStdin(reader))
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: suffix(`pipe %v`, dstpath),
+	})
+
+	// assert S3
+	assert.Assert(t, ensureS3Object(s3client, bucket, filename, content))
+}
+
+// pipe s3://bucket/prefix/target
+func TestUploadStdinToS3WithPrefix(t *testing.T) {
+	t.Parallel()
+
+	bucket := s3BucketFromTestName(t)
+
+	s3client, s5cmd := setup(t)
+
+	const (
+		filename = "testfile1.txt"
+		content  = "this is the content"
+	)
+
+	createBucket(t, s3client, bucket)
+
+	reader := bytes.NewBufferString(content)
+
+	dstpath := fmt.Sprintf("s3://%v/s5cmdtest/%v", bucket, filename)
+
+	cmd := s5cmd("pipe", dstpath)
+	result := icmd.RunCmd(cmd, icmd.WithStdin(reader))
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: equals(`pipe %v`, dstpath),
+	})
+
+	// assert s3 object
+	assert.Assert(t, ensureS3Object(s3client, bucket, fmt.Sprintf("s5cmdtest/%v", filename), content))
+}
+
+// pipe s3://bucket/object
+func TestUploadStdinToS3WithCustomName(t *testing.T) {
+	t.Parallel()
+
+	bucket := s3BucketFromTestName(t)
+
+	s3client, s5cmd := setup(t)
+
+	const (
+		filename = "testfile1.txt"
+		content  = "this is the content"
+	)
+
+	createBucket(t, s3client, bucket)
+
+	reader := bytes.NewBufferString(content)
+	dstpath := fmt.Sprintf("s3://%v/%v", bucket, filename)
+
+	cmd := s5cmd("pipe", dstpath)
+	result := icmd.RunCmd(cmd, icmd.WithStdin(reader))
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: equals(`pipe %v`, dstpath),
+	})
+
+	// assert s3 object
+	assert.Assert(t, ensureS3Object(s3client, bucket, filename, content))
+}
+
+// pipe -n s3://bucket/object
+func TestUploadStdinToS3WithNoClobber(t *testing.T) {
```

Comment:
How does this test work? I missed the `-n` assertions.

Reply:
Resolved. https://github.com/peak/s5cmd/pull/587/commits/c869dd1b9f071bf1d110ff5ee5078fdc9890ac49

Reply:
👍